### PR TITLE
Query optimization in CartRule

### PR DESCRIPTION
**Pull request by ghost**
Average improvement, from 3 ms, to 0.5 ms

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This change optimize a little bit the process of get the customer cart rules, the impact of improvement depends on the amount of vouchers.
| Type?         | improvement
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | no
| How to test?  | 

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/13585)

### CODE REVIEW

**Comment by Quetzacoalt91** [comment]
Hi there,

I'm afraid that without tests and any other proof that the behavior remains the same while the performance is improved, we won't be able to validate these changes.

**Comment by Rolige** [comment]
Hi @Quetzacoalt91,
Please don’t offend, but I don't think that's the best way to improve PrestaShop, just take a look at the queries, are made with **UNION ALL** to the same table, I guess it must be a very old query.

Regards!

**Comment by Quetzacoalt91** [comment]
> Please don’t offend, but I don't think that's the best way to improve PrestaShop

Actually, asking for a proof that the change works as before is a proper to improve the software. :)
There are a lot of pull-request to review, so asking for some tests before will accelerate the process.

**Comment by Rolige** [comment]
> There are a lot of pull-request to review, so asking for some tests before will accelerate the process.

I hope this video can help you: https://streamable.com/qt3cj

**Comment by PierreRambaud** [comment]
Ping @jocel1 Wdyt?

**Comment by PierreRambaud** [comment]
@Rolige Hi, don't forget this kind of tests must be done with a lot of cart rules, like 10, 100, 1000 & 10 000. :wink:

**Comment by jocel1** [comment]
@Rolige @PierreRambaud the OR condition should be slower than the UNION ALL on "old" MySQL version which are not able to optimize OR with an efficient index_merge :)
On which version of MySQL / MariaDB are you testing this?

**Comment by Rolige** [comment]
@jocel1 this is the thing, [according to system requirements for PrestaShop 1.7](https://devdocs.prestashop.com/1.7/basics/installation/system-requirements/), it works with MySQL 5.0+

Here are some screenshots of performance tests made, in spanish, sorry, but the difference in time is clear, which is the important part.

image
image
image
image
image
image
image
image
image
image
image

**Review by PierreRambaud — DISMISSED** [review]

**Comment by PierreRambaud** [comment]
I know I'm really late, could you please rebase your change :sweat_smile:

**Comment by Rolige** [comment]
> I know I'm really late, could you please rebase your change 😅

A rebase wasn't impossible, the code was modified recently in the PR https://github.com/PrestaShop/PrestaShop/pull/22518, we had to remove our previous commit and update it including the new changes on the mentioned PR, fortunately the change is minor, we have tested and its working fine, the branch is now updated!

File: classes/CartRule.php [hunk 1]
```diff
@@ -419,11 +415,19 @@ public static function getCustomerCartRules(
         }
 
         $sql_part1 = '* FROM `' . _DB_PREFIX_ . 'cart_rule` cr
-				LEFT JOIN `' . _DB_PREFIX_ . 'cart_rule_lang` crl ON (cr.`id_cart_rule` = crl.`id_cart_rule` AND crl.`id_lang` = ' . (int) $id_lang . ')';
+			LEFT JOIN `' . _DB_PREFIX_ . 'cart_rule_lang` crl ON (cr.`id_cart_rule` = crl.`id_cart_rule` AND crl.`id_lang` = ' . (int) $id_lang . ')';
+
+        $sql_where = ' WHERE ((cr.`id_customer` = ' . (int) $id_customer . ' OR (cr.`id_customer` = 0 AND (cr.`highlight` = 1 OR cr.`code` = ""))) OR cr.`group_restriction` = 1';
```
Comment (atomiix) [inline classes/CartRule.php]: I'm not sure about `OR cr.`group_restriction` = 1';` I think it has been removed on purpose in #16638

Comment (Rolige) [reply classes/CartRule.php]: You are right, we have updated the branch to keep this change introduced in https://github.com/PrestaShop/PrestaShop/pull/16638

**Review by atomiix — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]

**Comment by Robin-Fischer-PS** [comment]
Hello,

Since there is no issue linked, and no how to test, this should be tested by dev.

Thanks.

**Comment by atomiix** [comment]
@Rolige Can you fill the how to test section? I think it would be better if this PR is tested by the QA team

**Comment by Rolige** [comment]
> @Rolige Can you fill the how to test section? I think it would be better if this PR is tested by the QA team

Is complicated, because it's not a bug, and time to time we are doing this kind of optimization in which you only discover the problem in the real world and under very high traffic, as you can imagine, it's not easy to replicate these scenarios.

**Comment by PierreRambaud** [comment]
I didn't see any behaviour difference with or without this PR, seems to be faster on modern MySQL instances.

**Review by PierreRambaud — APPROVED** [review]

**Comment by PierreRambaud** [comment]
Thank you @Rolige

**Comment by jocel1** [comment]
Just do an EXPLAIN of the query plan to validate the optimizations, not just a SELECT, please!
Basically MySQL / MariaDB still sucks at optimizing OR queries (although there are been some improvements in the latest version with index_merge).

Try it on high concurrency workload and I guess you'll see the difference :)

**Pull request merged**